### ISSUE TITLE
IconsLayer starts flickering on constant changes

### ISSUE THREAD

**Issue by igor-dv** [uber/deck.gl#2869]
Hey all,

Sorry if I am asking an already answered question, couldn't find anything similar.

So, I am creating a POC for my work to use mapgox with deck.gl and one of the usecases I wanted to check is how the map behaves on constant objects' changes. Considering the reactive approach deck.gl implementing I suppose it's possible and I am doing something wrong. 

In our real world scenario we will have hundreds of moving objects that are being added and removed from the map. One of my final goals is to create a core map component that will be an abstraction over the 3rd party one and will expose some props api that we will use. Some part of the data will be stored in redux, then will be projected into the map's objects layer. 

For that purpose I've created a Storybook demo (placed [here](https://github.com/igor-dv/deck-gl-demo)), in that demo I am adding an Icon to the **IconsLayer** every second and at some point the map starts flickering... 

So basically my question is what I am doing wrong =)

Thanks.

**Comment by xintongxia** [comment]
One thing you can try to improve the performance is create the testLayers once, and update the layer prop `data` in your intervals.

```
  _renderLayers() {
    const {entities} = this.state;

    return new IconLayer({
      id: 'test',
      data: entities,

      pickable: true,
      wrapLongitude: true,

      getPosition: item => item.coordinates,
      getIcon: () => icon2,
      sizeScale: 100,
      getSize: 1,
    });
  }

  render() {
    const {mapStyle, viewPort} = this.state;

    return (
      <DeckGL
        controller
        width="100%"
        height="100%"
        initialViewState={viewPort}
        layers={this._renderLayers()}
      >
        ...
      </DeckGL>
    );
  }

```

**Comment by igor-dv** [comment]
How will this boost ? In your example, the `_renderLayers()` will be called every time the `render` happens.. 

I have changed my `componentDidMount` to this

```js
componentDidMount() {
        setInterval(() => {
            let { entities, iconLayer } = this.state;
    
            if (!entities) {
                entities = randomEntities(200);
            } else {
                entities = [...entities, ...randomEntities(1)];
            }

            if (!iconLayer) {
                iconLayer = new IconLayer({
                    id: 'test',
                    pickable: true,
                    wrapLongitude: true,
                    getPosition: item => item.coordinates,
                    data: entities,
                    getIcon: () => icon2,
                    sizeScale: 100,
                    getSize: 1,
                }); 
            } else {
                iconLayer.props.data = entities;
            }
    
            this.setState({
                entities,
                iconLayer,
                testLayes: [iconLayer], 
            });
        }, 1000);
    }
```
but nothing is rerendered now...

As far as I understand from the docs, layers are only the descriptors, so once there is something new in the layer I need to recreate the instance...

**Comment by igor-dv** [comment]
Seems like adding `requestAnimationFrame` into the interval helps.. but what if the changes are triggered by the store changes ? Are there any best practices of how to implement that ?

**Comment by Lowhearth** [comment]
The "Flickering" you are seeing is produced by the recalculation of the iconAtlas since you are using auto-packing with a version prior to the caching feature introduced by @xintongxia at #2728

Your package.json : 
>"dependencies": {
    "deck.gl": "^6.4.9",
    "mapbox-gl": "^0.53.1",
    "react": "^16.8.6",
    "react-dom": "^16.8.6",
    "react-map-gl": "^4.1.1",
    "react-scripts": "2.1.8"
  }

the flickering will not appear in v7.00 as long as you don't introduce new icons to the atlas.

**Comment by igor-dv** [comment]
@Lowhearth, thanks for the answer, didn't manage to check a new version yet. 

What do you mean by the "introduce new icons to the atlas" ? We are probably will be using sprites anyway, I can't think of a usecase we need to dynamically import an icon.

**Comment by Lowhearth** [comment]
I mean that the iconAtlas won't get resized and manipulated (which could cause flickering), as long as you don't provide new url's or iconIds that wasn't cached.
``` 
// extract icons from data
// return icons should be unique, and not cached or cached but url changed
export function getDiffIcons(data, getIcon, cachedIcons) {
  if (!data || !getIcon) {
    return null;
  }

  cachedIcons = cachedIcons || {};
  const icons = {};
  const {iterable, objectInfo} = createIterable(data);
  for (const object of iterable) {
    objectInfo.index++;
    const icon = getIcon(object, objectInfo);
    const id = getIconId(icon);

    if (!icon) {
      throw new Error('Icon is missing.');
    }

    if (!icon.url) {
      throw new Error('Icon url is missing.');
    }

    if (!icons[id] && (!cachedIcons[id] || icon.url !== cachedIcons[id].url)) {
      icons[id] = icon;
    }
  }
  return icons;
}
```

but if you provide your own spriteAtlas you wont have any problems since you will only load the texture once.
